server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java

```diff
@@ -106,31 +110,69 @@ public void getSegmentFiles(
             logger.trace("Downloading segments files from remote store {}", filesToFetch);
 
             RemoteSegmentMetadata remoteSegmentMetadata = remoteDirectory.readLatestMetadataFile();
-            List<StoreFileMetadata> downloadedSegments = new ArrayList<>();
+            List<StoreFileMetadata> toDownloadSegments = new ArrayList<>();
             Collection<String> directoryFiles = List.of(indexShard.store().directory().listAll());
             if (remoteSegmentMetadata != null) {
                 try {
                     indexShard.store().incRef();
                     indexShard.remoteStore().incRef();
                     final Directory storeDirectory = indexShard.store().directory();
+                    final ShardPath shardPath = indexShard.shardPath();
                     for (StoreFileMetadata fileMetadata : filesToFetch) {
                         String file = fileMetadata.name();
                         assert directoryFiles.contains(file) == false : "Local store already contains the file " + file;
-                        storeDirectory.copyFrom(remoteDirectory, file, file, IOContext.DEFAULT);
-                        downloadedSegments.add(fileMetadata);
+                        toDownloadSegments.add(fileMetadata);
                     }
-                    logger.trace("Downloaded segments from remote store {}", downloadedSegments);
+
+                    CountDownLatch countDownLatch = new CountDownLatch(1);
+                    LatchedActionListener<GetSegmentFilesResponse> completionListener = new LatchedActionListener<>(
+                        listener,
+                        countDownLatch
+                    );
+                    downloadSegments(storeDirectory, remoteDirectory, toDownloadSegments, shardPath, completionListener);
+                    countDownLatch.await();
+
+                    logger.trace("Downloaded segments from remote store {}", toDownloadSegments);
                 } finally {
                     indexShard.store().decRef();
                     indexShard.remoteStore().decRef();
                 }
             }
-            listener.onResponse(new GetSegmentFilesResponse(downloadedSegments));
         } catch (Exception e) {
             listener.onFailure(e);
         }
     }
 
+    private void downloadSegments(
+        Directory storeDirectory,
+        RemoteSegmentStoreDirectory remoteStoreDirectory,
+        List<StoreFileMetadata> toDownloadSegments,
+        ShardPath shardPath,
+        ActionListener<GetSegmentFilesResponse> completionListener
+    ) {
+
+        final AtomicInteger totalDownloadedSegments = new AtomicInteger(toDownloadSegments.size());
```

Comment:
Can you use the LatchedActionListener mechanism to avoid this separate counter? Pass in a countdownlatch with count equal to the number of segments to download and then just wait on that latch.

Reply:
Updated.

Reply:
Thanks for the idea. Updated the implementation with this approach.

Reply:
We don't need any latch here. I have commented below what we can do here.